File: implementations/python/packages/mcp-server/src/psp_cdl_mcp_server/revision.py
# SPDX-License-Identifier: Apache-2.0
"""Opt-in atomic callback selection and catalog publication, within one process."""
import hashlib
import re
import threading
import uuid
from types import SimpleNamespace
from psp_cdl_core import canonical_json
from psp_cdl_api_server import SecurityService, ServiceError
from psp_cdl_api_server.service import identifier
from psp_cdl_api_server.persistence import bounded, integer

REVISION_PROFILE="PSP-MCP-REVISION-0.1"
REVISION_KEY="psp-cdl.org/revision"
def fail(code,status=409): raise ServiceError(code,status)
def clone(value):
    try: return bounded(value)
    except Exception: fail("INVALID_REVISION_DATA",400)
def revision_digest(value): return hashlib.sha256(canonical_json(value).encode("utf-8")).hexdigest()
def identity(p): return canonical_json([p["tenantId"],p["subjectId"]])
def valid_catalog_revision(v):
    return type(v) is dict and set(v)=={"profile","epoch","generation","catalogDigest"} and v["profile"]==REVISION_PROFILE and identifier(v["epoch"]) and integer(v["generation"]) and v["generation"]>0 and type(v["catalogDigest"]) is str and re.fullmatch(r"[a-f0-9]{64}",v["catalogDigest"]) is not None

class RevisionedToolRegistry:
    def __init__(self,host,tools,epoch=None):
        self._epoch=str(uuid.uuid4()) if epoch is None else epoch
        if not identifier(self._epoch) or any(not callable(getattr(host,k,None)) for k in ("authenticate","authorize")): fail("INVALID_CONFIGURATION",400)
        self._host,self._auth=host,SecurityService(host)
        self._lock,self._active,self._generation=threading.Lock(),0,1
        self._state=self._prepare(tools)

    def _prepare(self,tools):
        if type(tools) is not list or len(tools)>1024: fail("INVALID_CONFIGURATION",400)
        handlers,catalog={},[]
        for tool in tools:
            if type(tool) is not dict: fail("INVALID_CONFIGURATION",400)
            invoke=tool.get("invoke")
            meta=clone({k:v for k,v in tool.items() if k!="invoke"})
            if set(meta)!={"name","revision","readOnly","inputSchema","outputSchema"} or type(meta["name"]) is not str or not re.fullmatch(r"[A-Za-z0-9_-]{1,64}",meta["name"]) or not identifier(meta["revision"]) or meta["readOnly"] is not True or any(type(meta[k]) is not dict or meta[k].get("type")!="object" for k in ("inputSchema","outputSchema")) or not callable(invoke) or meta["name"] in handlers: fail("INVALID_CONFIGURATION",400)
            handlers[meta["name"]]=invoke
            catalog.append({"name":meta["name"],"inputSchema":meta["inputSchema"],"outputSchema":meta["outputSchema"],"_meta":{REVISION_KEY:{"toolRevision":meta["revision"]}}})
        catalog.sort(key=lambda t:t["name"])
        return {"handlers":handlers,"catalog":clone(catalog),"digest":revision_digest(catalog)}

    def _revision(self): return {"profile":REVISION_PROFILE,"epoch":self._epoch,"generation":self._generation,"catalogDigest":self._state["digest"]}
    @property
    def revision(self):
        with self._lock: return self._revision()

    def publish(self,expected,tools):
        with self._lock:
            if not valid_catalog_revision(expected) or canonical_json(expected)!=canonical_json(self._revision()): fail("REVISION_CONFLICT")
            if self._active: fail("REGISTRY_BUSY")
            if self._generation==9_007_199_254_740_991: fail("REVISION_EXHAUSTED")
            next_state=self._prepare(tools)
            self._state=next_state
            self._generation+=1
            return self._revision()

    def _authorized(self,token,expected,name,phase):
        p=self._auth.authenticate(token)
        if identity(p)!=identity(expected) or ("tools:list" if phase=="discover" else "tools:call") not in p["scopes"] or self._host.authorize(clone(p),name,phase) is not True: fail("FORBIDDEN",403)
        return p

    def service(self,cancelled=lambda:False):
        def check():
            if cancelled() is not False: fail("CANCELLED")
        def discover(token,p):
            self._authorized(token,p,None,"discover")
            check()
            with self._lock: return clone({"tools":self._state["catalog"],"_meta":{REVISION_KEY:self._revision()}})
        def call(name,args,token,p,raw):
            data,pre=clone(args),clone(raw)
            if type(data) is not dict or type(pre) is not dict or set(pre)!={"profile","epoch","generation","catalogDigest","toolRevision","inputDigest"}: fail("INVALID_PRECONDITION",400)
            principal=self._authorized(token,p,name,"invoke")
            check()
            with self._lock:
                tool=next((t for t in self._state["catalog"] if t["name"]==name),None)
                expected={**self._revision(),"toolRevision":tool["_meta"][REVISION_KEY]["toolRevision"] if tool else None,"inputDigest":revision_digest(data)}
                if tool is None or canonical_json(pre)!=canonical_json(expected): fail("REVISION_MISMATCH")
                invoke=self._state["handlers"][name]
                self._active+=1
            try:
                output=clone(invoke(data,{"principal":clone(principal),"cancelled":cancelled}))
                if type(output) is not dict: fail("INVALID_OUTPUT",500)
                check()
                self._authorized(token,p,name,"release")
                check()
                return clone({"data":output,"meta":{REVISION_KEY:pre}})
            finally:
                with self._lock: self._active-=1
        return SimpleNamespace(authenticate=self._auth.authenticate,discover=lambda *_:fail("REVISION_REQUIRED"),call_tool=lambda *_:fail("REVISION_REQUIRED"),revisions=SimpleNamespace(profile=REVISION_PROFILE,discover=discover,call_tool=call))

File: implementations/typescript/packages/mcp-server/src/revision.ts
// SPDX-License-Identifier: Apache-2.0
import {createHash,randomUUID} from "node:crypto";
import {canonicalJson,record} from "@psp-cdl/core";
import {SecurityService,ServiceError,identifier,type Principal} from "@psp-cdl/api-server";
import {bounded} from "@psp-cdl/api-server/persistence";
import type {McpToolService} from "./server.js";

export const REVISION_PROFILE="PSP-MCP-REVISION-0.1";
export const REVISION_KEY="psp-cdl.org/revision";
export interface CatalogRevision {profile:typeof REVISION_PROFILE;epoch:string;generation:number;catalogDigest:string}
export interface RevisionPrecondition extends CatalogRevision {toolRevision:string;inputDigest:string}
export interface RevisionTool {
  name:string;revision:string;readOnly:true;inputSchema:Record<string,unknown>;outputSchema:Record<string,unknown>;
  invoke(args:Record<string,unknown>,context:{principal:Principal;cancelled:()=>boolean}):Record<string,unknown>|Promise<Record<string,unknown>>;
}
export interface RevisionHost {
  authenticate(token:string):Principal|null|Promise<Principal|null>;
  authorize(principal:Principal,name:string|null,phase:"discover"|"invoke"|"release"):boolean|Promise<boolean>;
}
export interface RevisionService {
  profile:typeof REVISION_PROFILE;
  discover(token:string,principal:Principal):Promise<Record<string,unknown>>;
  callTool(name:string,args:Record<string,unknown>,token:string,principal:Principal,precondition:unknown):Promise<{data:Record<string,unknown>;meta:Record<string,unknown>}>;
}
export const revisionDigest=(value:unknown):string=>createHash("sha256").update(canonicalJson(value)).digest("hex");
function fail(code:string,status=409):never {throw new ServiceError(code,status);}
const clone=(v:unknown):any=>{try{return bounded(v);}catch{return fail("INVALID_REVISION_DATA",400);}};
const identity=(p:Principal)=>canonicalJson([p.tenantId,p.subjectId]);
export function validCatalogRevision(v:unknown):v is CatalogRevision {
  return record(v)&&Object.keys(v).sort().join(",")==="catalogDigest,epoch,generation,profile"&&v.profile===REVISION_PROFILE&&identifier(v.epoch)&&typeof v.generation==="number"&&Number.isSafeInteger(v.generation)&&v.generation>0&&typeof v.catalogDigest==="string"&&v.catalogDigest.length===64&&/^[a-f0-9]{64}$/.test(v.catalogDigest);
}

/** All publishers and invocations for this catalog must share this process-local registry. */
export class RevisionedToolRegistry {
  private readonly auth:SecurityService;
  private active=0;
  private generation=1;
  private state:ReturnType<RevisionedToolRegistry["prepare"]>;
  constructor(private readonly host:RevisionHost,tools:RevisionTool[],private readonly epoch:string=randomUUID()) {
    if(!identifier(epoch)||typeof host.authenticate!=="function"||typeof host.authorize!=="function")fail("INVALID_CONFIGURATION",400);
    this.auth=new SecurityService({authenticate:t=>host.authenticate(t),resolve:()=>null,now:Date.now});
    this.state=this.prepare(tools);
  }
  private prepare(tools:RevisionTool[]) {
    if(!Array.isArray(tools)||tools.length>1024)fail("INVALID_CONFIGURATION",400);
    const handlers=new Map<string,RevisionTool["invoke"]>(),catalog:Record<string,unknown>[]=[];
    for(const tool of tools) {
      if(!record(tool))fail("INVALID_CONFIGURATION",400);
      const {invoke,...value}=tool,meta=clone(value);
      if(Object.keys(meta).sort().join(",")!=="inputSchema,name,outputSchema,readOnly,revision"||typeof meta.name!=="string"||!/^[A-Za-z0-9_-]{1,64}$/.test(meta.name)||/[^A-Za-z0-9_-]/.test(meta.name)||!identifier(meta.revision)||meta.readOnly!==true||!record(meta.inputSchema)||meta.inputSchema.type!=="object"||!record(meta.outputSchema)||meta.outputSchema.type!=="object"||typeof invoke!=="function"||handlers.has(meta.name))fail("INVALID_CONFIGURATION",400);
      handlers.set(meta.name,invoke);
      catalog.push({name:meta.name,inputSchema:meta.inputSchema,outputSchema:meta.outputSchema,_meta:{[REVISION_KEY]:{toolRevision:meta.revision}}});
    }
    catalog.sort((a,b)=>(a.name as string)<(b.name as string)?-1:1);
    return {handlers,catalog:clone(catalog) as Record<string,unknown>[],digest:revisionDigest(catalog)};
  }
  get revision():CatalogRevision {return {profile:REVISION_PROFILE,epoch:this.epoch,generation:this.generation,catalogDigest:this.state.digest};}
  /** Fail-fast compare-and-replace. A lease prevents replacement until buffered completion. */
  publish(expected:CatalogRevision,tools:RevisionTool[]):CatalogRevision {
    if(!validCatalogRevision(expected)||canonicalJson(expected)!==canonicalJson(this.revision))fail("REVISION_CONFLICT");
    if(this.active)fail("REGISTRY_BUSY");
    if(this.generation===Number.MAX_SAFE_INTEGER)fail("REVISION_EXHAUSTED");
    const next=this.prepare(tools);this.state=next;this.generation++;
    return this.revision;
  }
  private async authorized(token:string,expected:Principal,name:string|null,phase:"discover"|"invoke"|"release"):Promise<Principal> {
    const p=await this.auth.authenticate(token);
    if(identity(p)!==identity(expected)||!p.scopes.includes(phase==="discover"?"tools:list":"tools:call")||await this.host.authorize(clone(p),name,phase)!==true)fail("FORBIDDEN",403);
    return p;
  }
  service(cancelled:()=>boolean=()=>false):McpToolService {
    const check=()=>{if(cancelled()!==false)fail("CANCELLED");};
    return {
      authenticate:t=>this.auth.authenticate(t),
      discover:()=>fail("REVISION_REQUIRED"),callTool:()=>fail("REVISION_REQUIRED"),
      revisions:{profile:REVISION_PROFILE,
        discover:async(token,p)=>{
          await this.authorized(token,p,null,"discover");check();
          return clone({tools:this.state.catalog,_meta:{[REVISION_KEY]:this.revision}});
        },
        callTool:async(name,args,token,p,raw)=>{
          const data=clone(args),pre=clone(raw);
          if(!record(data)||!record(pre)||Object.keys(pre).sort().join(",")!=="catalogDigest,epoch,generation,inputDigest,profile,toolRevision")fail("INVALID_PRECONDITION",400);
          const principal=await this.authorized(token,p,name,"invoke");check();
          const tool=this.state.catalog.find(t=>t.name===name),revision=this.revision;
          const expected={...revision,toolRevision:(tool?._meta as any)?.[REVISION_KEY]?.toolRevision,inputDigest:revisionDigest(data)};
          if(!tool||canonicalJson(pre)!==canonicalJson(expected))fail("REVISION_MISMATCH");
          const invoke=this.state.handlers.get(name)!;
          // No await between comparison, selecting the immutable callback and acquiring its lease.
          this.active++;
          try {
            const output=clone(await invoke(data,{principal:clone(principal),cancelled}));
            if(!record(output))fail("INVALID_OUTPUT",500);
            check();await this.authorized(token,p,name,"release");check();
            return clone({data:output,meta:{[REVISION_KEY]:pre}});
          }finally{this.active--;}
        }
      }
    };
  }
}
